### PR TITLE
feat/35 graceful error handling

### PR DESCRIPTION
**Pull request by jeremy-babylonchain**
#35 

- display error cause
- gracefully handle errors from different API calls with a modal popup. allow user to close modal or retry the failed function manually by clicking "try again"

### CODE REVIEW

File: src/app/api/getDelegations.ts
Comment (jrwbabylonchain) [inline src/app/api/getDelegations.ts:1]: I think u need to fix the linting. `npm run format:fix`

**Comment by jrwbabylonchain** [comment]
I'm not an expert on react. i heard something about react error boundary which seems to be a global error handler. https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary
Is this something we can do as well?

**Comment by gbarkhatov** [comment]
@jrwbabylonchain the usage of error boundaries is usually to have a fallback, i.e. when things crash on your component, you provide something with the "proper" state

This above ticket is more about showing errors inside the popup/modal that X went wrong

File: src/app/components/Delegations/Delegations.tsx [hunk 1]
```diff
@@ -172,6 +175,14 @@ export const Delegations: React.FC<DelegationsProps> = ({
       handleUnbond(id);
     } catch (error: Error | any) {
       console.error(error?.message || error);
```
Comment (gbarkhatov) [inline src/app/components/Delegations/Delegations.tsx:177]: Should we remove console.error?

File: src/app/page.tsx [hunk 2]
```diff
@@ -158,6 +237,14 @@ const Home: React.FC<HomeProps> = () => {
       setPublicKeyNoCoord(publicKeyNoCoord.toString("hex"));
     } catch (error: Error | any) {
       console.error(error?.message || error);
```
Comment (gbarkhatov) [inline src/app/page.tsx:239]: Remove? So as everywhere from the app (3 times in page.tsx, 1 time in Delegations)

File: src/app/api/getStakers.ts
Comment (gbarkhatov) [inline src/app/api/getStakers.ts:1]: remove encode?

File: src/app/components/Modals/ErrorModal.tsx [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  errorState?: ErrorState;
+  errorTime: Date;
+  errorMetadata?: ErrorMetadata;
+}
+
+export const ErrorModal: React.FC<ErrorModalProps> = ({
+  open,
+  onClose,
+  onRetry,
+  errorCode,
```
Comment (gbarkhatov) [inline src/app/components/Modals/ErrorModal.tsx:23]: `errorCode` is not used, let's remove

File: src/app/components/Modals/ErrorModal.tsx [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+  errorMessage,
+  errorState,
+  errorTime,
+  errorMetadata,
```
Comment (gbarkhatov) [inline src/app/components/Modals/ErrorModal.tsx:27]: same here, `errorMetadata` is not used

File: src/app/components/Modals/ErrorModal.tsx
Comment (gbarkhatov) [inline src/app/components/Modals/ErrorModal.tsx:1]: I like this implementation 👍

File: src/app/components/Stakers/Stakers.tsx [hunk 5]
```diff
@@ -4,17 +4,41 @@ import { Tooltip } from "react-tooltip";
 
 import { getStakers } from "@/app/api/getStakers";
 import { Staker } from "./Staker";
+import { useError } from "@/app/context/Error/ErrorContext";
+import { useEffect } from "react";
```
Comment (gbarkhatov) [inline src/app/components/Stakers/Stakers.tsx:8]: Please move global imports above

File: src/app/components/Stakers/Stakers.tsx
Comment (gbarkhatov) [inline src/app/components/Stakers/Stakers.tsx:1]: There's a TODO on line 76:
// TODO handle API data errors with a proper error boundary
Please remove it as your PR is solving the issue

File: src/app/context/Error/ErrorContext.tsx [hunk 6]
```diff
@@ -0,0 +1,79 @@
+import { ErrorState, ErrorType } from "@/app/types/errorState";
+import React, {
+  createContext,
+  useContext,
+  useState,
+  useCallback,
+  ReactNode,
+} from "react";
```
Comment (gbarkhatov) [inline src/app/context/Error/ErrorContext.tsx:8]: Please use
```
global imports

local imports
```

I.e. 
```ts
import React, {
  createContext,
  useContext,
  useState,
  useCallback,
  ReactNode,
} from "react";

import { ErrorState, ErrorType } from "@/app/types/errorState";
```

Also, `ErrorState` is not used

File: src/app/page.tsx [hunk 7]
```diff
@@ -54,56 +57,132 @@ const Home: React.FC<HomeProps> = () => {
...
+    refetch: refetchFinalityProvidersData,
+  } = useQuery({
     queryKey: ["finality providers"],
     queryFn: getFinalityProviders,
     refetchInterval: 60000, // 1 minute
     select: (data) => data.data,
+    retry: false,
+  });
+
+  useEffect(() => {
```
Comment (gbarkhatov) [inline src/app/page.tsx:116]: Is this useEffect is okay for our performance?

File: src/app/providers.tsx [hunk 8]
```diff
@@ -5,14 +5,17 @@ import { QueryClientProvider, QueryClient } from "@tanstack/react-query";
 import { ReactQueryDevtools } from "@tanstack/react-query-devtools";
 import { ReactQueryStreamedHydration } from "@tanstack/react-query-next-experimental";
 import { ThemeProvider } from "next-themes";
+import { ErrorProvider } from "./context/Error/ErrorContext";
```
Comment (gbarkhatov) [inline src/app/providers.tsx:8]: Please separate global and local imports with one empty line

File: src/app/types/errorState.ts [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+  GET_STAKING = "GET_STAKING",
+  GET_DELEGATION = "GET_DELEGATION",
+  GET_FINALITY_PROVIDER = "GET_FINALITY_PROVIDER",
+  GET_GLOBAL_PARAMS = "GET_GLOBAL_PARAMS",
+  GET_STATS = "GET_STATS",
+  GET_UNBOUNDING_ELIGIBILITY = "GET_UNBOUNDING_ELIGIBILITY",
+  POST_UNBOUNDING = "POST_UNBOUNDING",
+  SWITCH_NETWORK = "SWITCH_NETWORK",
+  WITHDRAW = "WITHDRAW",
+}
```
Comment (gbarkhatov) [inline src/app/types/errorState.ts:11]: Let's add a STAKING error and try to change some of the parameters in `handleSign` so you will receive an error when creating staking transaction

**Comment by gbarkhatov** [comment]
@jeremy-babylonchain also please run `npm run format:fix`

**Comment by jeremy-babylonchain** [comment]
> Does the `retry` suppose to do anything? when i tested it locally, it didn't perform any action
> 
>  Screen.Recording.2024-05-20.at.11.08.54.AM.mov

Resolved this issue

**Comment by jeremy-babylonchain** [comment]
Staking error desktop
CleanShot 2024-05-21 at 00 00 52

Staking error mobile

https://github.com/babylonchain/simple-staking/assets/168515712/c54858b1-1baa-40e6-8289-0c67e5a71e55

Connection error desktop
CleanShot 2024-05-20 at 23 50 04

Connection error mobile
CleanShot 2024-05-20 at 23 51 07

File: src/app/page.tsx [hunk 10]
```diff
@@ -41,54 +44,141 @@ const Home: React.FC<HomeProps> = () => {
...
+      refetchFunction: refetchGlobalParamsVersion,
+    });
+  }, [
+    hasFinalityProvidersError,
+    hasGlobalParamsVersionError,
+    hasDelegationsError,
+    hasStatsError,
+    isRefetchFinalityProvidersError,
+  ]);
+
```
Comment (gbarkhatov) [inline src/app/page.tsx:181]: Taptoor below should be Taproot

File: src/app/page.tsx [hunk 11]
```diff
@@ -142,7 +232,14 @@ const Home: React.FC<HomeProps> = () => {
       setAddress(address);
       setPublicKeyNoCoord(publicKeyNoCoord.toString("hex"));
     } catch (error: Error | any) {
-      console.error(error?.message || error);
+      showError({
+        error: {
+          message: error.message,
+          errorState: ErrorState.SWITCH_NETWORK,
```
Comment (gbarkhatov) [inline src/app/page.tsx:238]: Maybe we use `ErrorState.WALLET`?

File: src/app/components/Stakers/Stakers.tsx (hunk 5 above)
Comment (gbarkhatov) [reply src/app/components/Stakers/Stakers.tsx:8]: @jeremy-babylonchain move `react` import above
```
global imports
(empty line)
alias imports (with @)
local imports (with dos)
```

**Review by gbarkhatov — APPROVED** [review]
Looks good to me, @jeremy-babylonchain please resolve minor comments and double check that you don't overwrite the previous changes from the `main` branch

@jrwbabylonchain as soon as you approve this can be merged

File: src/app/components/Modals/ErrorModal.tsx [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+  };
+
+  const getErrorTitle = () => {
+    switch (errorState) {
+      case ErrorState.GET_STAKING:
+      case ErrorState.GET_DELEGATION:
+      case ErrorState.GET_STATS:
+      case ErrorState.GET_FINALITY_PROVIDER:
+      case ErrorState.GET_GLOBAL_PARAMS:
+      case ErrorState.GET_UNBOUNDING_ELIGIBILITY:
```
Comment (jrwbabylonchain) [inline src/app/components/Modals/ErrorModal.tsx:58]: nit: IMO, all those state are the same state. i.e `SERVER_ERROR`

**Review by jrwbabylonchain — APPROVED** [review]

**Pull request merged**